junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ConcurrentHierarchicalTestExecutorService.java

```diff
@@ -226,13 +227,25 @@ void processQueueEntries(WorkerLease workerLease, BooleanSupplier doneCondition)
 					LOGGER.trace(() -> "yielding resource lock");
 					break;
 				}
-				var entry = workQueue.poll();
-				if (entry == null) {
-					LOGGER.trace(() -> "no queue entry available");
+				if (workQueue.isEmpty()) {
+					LOGGER.trace(() -> "no queue entries available");
 					break;
 				}
-				LOGGER.trace(() -> "processing: " + entry.task);
-				execute(entry);
+				processQueueEntries();
+			}
+		}
+
+		private void processQueueEntries() {
+			var queueEntriesByResult = tryToStealWorkWithoutBlocking(workQueue);
+			var queueModified = queueEntriesByResult.containsKey(WorkStealResult.EXECUTED_BY_THIS_WORKER) //
+					|| queueEntriesByResult.containsKey(WorkStealResult.EXECUTED_BY_DIFFERENT_WORKER);
+			if (queueModified) {
+				return;
+			}
+			var entriesRequiringResourceLocks = queueEntriesByResult.get(WorkStealResult.RESOURCE_LOCK_UNAVAILABLE);
+			if (entriesRequiringResourceLocks != null) {
+				// One entry at a time to avoid blocking too much
+				tryToStealWork(entriesRequiringResourceLocks.get(0), BlockingMode.BLOCKING);
```

Comment:
Could you please elaborate on this? Is there evidence this is better than doing all of them?